src/combinatorics/simple_graph/ends/defs.lean

```diff
@@ -0,0 +1,256 @@
+-/
+lemma mem_of_adj : ∀ {C : G.comp_out K} (c d : V), c ∈ C → d ∉ K → G.adj c d → d ∈ C :=
+λ C c d ⟨cnK,h⟩ dnK cd,
+  ⟨ dnK, by { rw [←h, connected_component.eq], exact adj.reachable cd.symm, } ⟩
+
+/--
+Assuming `G` is preconnected and `K` not empty, given any connected component `C` outside of `K`,
+there exists a vertex `k ∈ K` adjacent to a vertex `v ∈ C`.
+-/
+lemma exists_adj_boundary_pair (Gc : G.preconnected) (hK : K.nonempty) :
+  ∀ (C : G.comp_out K), ∃ (ck : V × V), ck.1 ∈ C ∧ ck.2 ∈ K ∧ G.adj ck.1 ck.2 :=
+begin
+  refine comp_out.ind (λ v vnK, _),
+  let C : G.comp_out K := G.comp_out_mk vnK,
+  let dis := set.disjoint_iff.mp C.disjoint_right,
+  by_contra' h,
+  suffices : set.univ = (C : set V),
+  { exact dis ⟨hK.some_spec, this ▸ (set.mem_univ hK.some)⟩, },
+  symmetry,
+  rw set.eq_univ_iff_forall,
+  rintro u,
+  by_contradiction unC,
+  obtain ⟨p⟩ := Gc v u,
+  obtain ⟨⟨⟨x, y⟩, xy⟩, d, xC, ynC⟩ :=
+    p.exists_boundary_dart (C : set V) (G.comp_out_mk_mem vnK) unC,
+  exact ynC (mem_of_adj x y xC (λ (yK : y ∈ K), h ⟨x, y⟩ xC yK xy) xy),
+end
+
+/--
+If `K ⊆ L`, the components outside of `L` are all contained in a single component outside of `K`.
+-/
+@[reducible] def hom (h : K ⊆ L) (C : G.comp_out L) : G.comp_out K := C.map (G.out_hom h)
+
+lemma subset_hom (C : G.comp_out L) (h : K ⊆ L) : (C : set V) ⊆ (C.hom h : set V) := by
+{ rintro c ⟨cL,rfl⟩, exact ⟨λ h', cL (h h'), rfl⟩ }
+
+lemma _root_.simple_graph.comp_out_mk_mem_hom (G : simple_graph V) {v : V} (vK : v ∈ Kᶜ)
+  (h : L ⊆ K) : v ∈ (G.comp_out_mk vK).hom h :=
+subset_hom (G.comp_out_mk vK) h (G.comp_out_mk_mem vK)
+
+lemma hom_eq_iff_le (C : G.comp_out L) (h : K ⊆ L) (D : G.comp_out K) :
+  C.hom h = D ↔ (C : set V) ⊆ (D : set V) :=
+⟨ λ h', h' ▸ (C.subset_hom h), C.ind (λ v vnL vD, (vD ⟨vnL, rfl⟩).some_spec) ⟩
+
+lemma hom_eq_iff_not_disjoint (C : G.comp_out L) (h : K ⊆ L) (D : G.comp_out K) :
+  C.hom h = D ↔ ¬ disjoint (C : set V) (D : set V) :=
+begin
+  rw set.not_disjoint_iff,
+  split,
+  { rintro rfl,
+    apply C.ind (λ x xnL, _),
+    exact ⟨x, ⟨xnL,rfl⟩, ⟨(λ xK, xnL (h xK)), rfl⟩⟩, },
+  { apply C.ind (λ x xnL, _),
+    rintro ⟨x,⟨_,e₁⟩,⟨_,rfl⟩⟩,
+    rw ←e₁, refl, },
+end
+
+lemma hom_refl (C : G.comp_out L) : C.hom (subset_refl L) = C :=
+by { change C.map _ = C, rw [G.out_hom_refl L, C.map_id], }
+
+lemma hom_trans (C : G.comp_out L) (h : K ⊆ L) (h' : M ⊆ K) :
+  C.hom (h'.trans h) = (C.hom h).hom h' :=
+by { change C.map _ = (C.map _).map _, rw [G.out_hom_trans, C.map_comp], }
+
+lemma hom_mk {v : V} (vnL : v ∉ L) (h : K ⊆ L) :
+  (G.comp_out_mk vnL).hom h = (G.comp_out_mk (set.not_mem_subset h vnL)) := rfl
+
+lemma hom_inf (C : G.comp_out L) (h : K ⊆ L) (Cinf : (C : set V).infinite) :
+  (C.hom h : set V).infinite := set.infinite.mono (C.subset_hom h) Cinf
+
+lemma inf_iff_in_all_ranges {K : finset V} (C : G.comp_out K) :
+  (C : set V).infinite ↔ ∀ L (h : K ⊆ L), ∃ D : G.comp_out L, C = D.hom h :=
```

Comment:
Good idea. At first we had two functors: the general one with values all connected components, and the subfunctor of infinite connected components, but I figured it would be easier to only work with the one functor (fewer defs, more lemmas).
But then since the functor itself isn't "surjective" (as in the Mittag-Leffler defs), I'm wondering how more messy it will be to show things "If there exists some `K : finset` such that no `K ≤ L : finset` has more infinite components than K, then there are at most `card (G.comp_out K)` ends".

Reply:
I think the subfunctor of infinite connected components is [category_theory.functor.to_eventual_ranges](https://leanprover-community.github.io/mathlib_docs/category_theory/mittag_leffler.html#category_theory.functor.to_eventual_ranges) of the current functor, right? We already have an equivalence to switch between the sections of both, so I wouldn't expect this to be too hard.